modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/SnapshotCheckProcess.java

```diff
@@ -0,0 +1,475 @@
+
+        // Local meta might be null if current node started after the snapshot creation or placement.
+        if (!req.nodes.contains(kctx.localNodeId()) || ctx.locMeta == null)
+            ctx.fut.onDone();
+        else {
+            File snpDir = kctx.cache().context().snapshotMgr().snapshotLocalDir(req.snapshotName(), req.snapshotPath());
+
+            kctx.cache().context().snapshotMgr().checker().checkPartitions(ctx.locMeta, snpDir, req.groups(), false, true, false)
+                .whenComplete((res, err) -> {
+                    if (err != null)
+                        ctx.fut.onDone(err);
+                    else
+                        ctx.fut.onDone(new CheckResultDTO(res));
+                });
+        }
+
+        return ctx.fut;
+    }
+    
+    /** */
+    private Map<ClusterNode, Exception> collectErrors(@Nullable Map<UUID, Throwable> errors, @Nullable Set<UUID> requiredNodes) {
+        if (F.isEmpty(errors))
+            return Collections.emptyMap();
+
+        return errors.entrySet().stream()
+            .filter(e -> (requiredNodes == null || requiredNodes.contains(e.getKey())) && e.getValue() != null)
+            .collect(Collectors.toMap(e -> kctx.cluster().get().node(e.getKey()), e -> asException(e.getValue())));
+    }
+
+    /** */
+    private Map<ClusterNode, Map<PartitionKeyV2, PartitionHashRecordV2>> collectPartsHashes(
+        @Nullable Map<UUID, CheckResultDTO> results,
+        Collection<UUID> requiredNodes
+    ) {
+        if (F.isEmpty(results))
+            return Collections.emptyMap();
+
+        return results.entrySet().stream()
+            .filter(e -> requiredNodes.contains(e.getKey()) && e.getValue() != null)
+            .collect(Collectors.toMap(e -> kctx.cluster().get().node(e.getKey()), e -> e.getValue().partsHashes));
+    }
+
+    /**
+     * @param snpName Snapshot name of the validation process. If {@code null}, ignored.
+     * @param reqId  If {@code snpName} is {@code null}, is used to find the operation request.
+     * @return Current snapshot checking context by {@code snpName} or {@code reqId}.
+     */
+    private @Nullable SnapshotCheckContext context(@Nullable String snpName, UUID reqId) {
+        return snpName == null
+            ? contexts.values().stream().filter(ctx -> ctx.req.reqId.equals(reqId)).findFirst().orElse(null)
+            : contexts.get(snpName);
+    }
+
+    /** Phase 1 beginning: prepare, collect and check local metas. */
+    private IgniteInternalFuture<CheckResultDTO> prepareAndCheckMetas(SnapshotCheckProcessRequest req) {
+        SnapshotCheckContext ctx = contexts.computeIfAbsent(req.snapshotName(), snpName -> new SnapshotCheckContext(req));
+
+        if (!ctx.req.equals(req)) {
+            return new GridFinishedFuture<>(new IllegalStateException("Validation of snapshot '" + req.snapshotName()
+                + "' has already started. Request=" + ctx + '.'));
+        }
+
+        if (!req.nodes.contains(kctx.localNodeId())) {
```

Comment:
Can we check it before creating the context? If node should be skipped then we can just return `new GridFinishedFuture()`.